File: modulo-01-fundamentos-arquitectura/codigo/src/main/java/pe/joedayz/microservicios/modulo01/ddd/shared/Money.java
package pe.joedayz.microservicios.modulo01.ddd.shared;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Objects;

/**
 * Value Object para dinero: monto + moneda.
 *
 * <p>Caracteristicas de un Value Object bien hecho:
 * <ul>
 *   <li><b>Inmutable</b>: las operaciones devuelven nuevos {@code Money}.</li>
 *   <li><b>Auto-validado</b>: no puede existir en estado invalido (moneda nula, escala mala).</li>
 *   <li><b>Igualdad por valor</b>: S/100 PEN siempre es igual a otro S/100 PEN.</li>
 *   <li><b>Sin identidad</b>: no tiene id.</li>
 * </ul>
 *
 * <p>Nunca uses {@code double} para dinero: usa {@link BigDecimal} para evitar errores
 * de redondeo binario.
 */
public record Money(BigDecimal amount, String currency) {

    public Money {
        Objects.requireNonNull(amount, "amount no puede ser null");
        Objects.requireNonNull(currency, "currency no puede ser null");
        if (currency.length() != 3) {
            throw new IllegalArgumentException("currency debe ser ISO-4217 de 3 letras, ej. PEN/USD");
        }
        amount = amount.setScale(2, RoundingMode.HALF_EVEN);
    }

    public static Money of(String amount, String currency) {
        return new Money(new BigDecimal(amount), currency);
    }

    public static Money zero(String currency) {
        return new Money(BigDecimal.ZERO, currency);
    }

    public Money plus(Money other) {
        requireSameCurrency(other);
        return new Money(this.amount.add(other.amount), this.currency);
    }

    public Money times(int factor) {
        return new Money(this.amount.multiply(BigDecimal.valueOf(factor)), this.currency);
    }

    public boolean isGreaterThan(Money other) {
        requireSameCurrency(other);
        return this.amount.compareTo(other.amount) > 0;
    }

    private void requireSameCurrency(Money other) {
        if (!this.currency.equals(other.currency)) {
            throw new IllegalArgumentException(
                    "No se pueden operar monedas distintas: " + this.currency + " vs " + other.currency);
        }
    }

    @Override
    public String toString() {
        return currency + " " + amount;
    }
}

File: modulo-01-fundamentos-arquitectura/codigo-dotnet/src/Modulo01.Fundamentos/Ddd/Shared/Money.cs
namespace JoeDayz.Microservicios.Modulo01.Ddd.Shared;

/// <summary>
/// Value object de dinero. Inmutable, se compara por valor y protege sus invariantes:
/// moneda ISO-4217 de 3 letras y redondeo bancario a 2 decimales.
/// En .NET usamos <see cref="decimal"/> (equivalente a BigDecimal para importes).
/// </summary>
public sealed record Money
{
    public Money(decimal amount, string currency)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(currency, nameof(currency));
        if (currency.Length != 3)
        {
            throw new ArgumentException("currency debe ser ISO-4217 de 3 letras, ej. PEN/USD", nameof(currency));
        }

        Amount = Math.Round(amount, 2, MidpointRounding.ToEven);
        Currency = currency;
    }

    public decimal Amount { get; }

    public string Currency { get; }

    public static Money Of(decimal amount, string currency) => new(amount, currency);

    public static Money Zero(string currency) => new(0m, currency);

    public Money Plus(Money other)
    {
        RequireSameCurrency(other);
        return new Money(Amount + other.Amount, Currency);
    }

    public Money Minus(Money other)
    {
        RequireSameCurrency(other);
        return new Money(Amount - other.Amount, Currency);
    }

    public Money Times(int factor) => new(Amount * factor, Currency);

    public bool IsGreaterThan(Money other)
    {
        RequireSameCurrency(other);
        return Amount > other.Amount;
    }

    public bool IsNegative => Amount < 0m;

    private void RequireSameCurrency(Money other)
    {
        if (Currency != other.Currency)
        {
            throw new ArgumentException(
                $"No se pueden operar monedas distintas: {Currency} vs {other.Currency}", nameof(other));
        }
    }

    public override string ToString() => $"{Currency} {Amount:0.00}";
}
